src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1998,61 +1997,63 @@ void GenerateResetMessage()
         /// </summary>
         internal async Task Reset(bool async, CancellationToken cancellationToken = default)
         {
-            Debug.Assert(IsReady);
-
-            // Our buffer may contain unsent prepended messages, so clear it out.
-            // In practice, this is (currently) only done when beginning a transaction or a transaction savepoint.
-            WriteBuffer.Clear();
-            PendingPrependedResponses = 0;
-
-            ResetReadBuffer();
-
-            Transaction?.UnbindIfNecessary();
-
             bool endBindingScope;
 
-            // Must rollback transaction before sending DISCARD ALL
-            switch (TransactionStatus)
+            // We start user action as a guard against keepalive
+            using (StartUserAction(cancellationToken))
```

Comment:
Unrelated comment - unless I'm missing something, cancellationToken never seems to be set for the entire Close code path. This is generally by design (dispose/close generally accept no cancellation tokens in .NET), we could clean this up and remove the parameter.